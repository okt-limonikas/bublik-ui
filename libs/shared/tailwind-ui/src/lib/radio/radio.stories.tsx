/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { StoryFn, Meta } from '@storybook/react';

import { withBackground } from '../storybook-bg';
import { Radio, RadioProps } from './radio';

export default {
	component: Radio,
	title: 'form/Radio',
	decorators: [withBackground]
} as Meta;

export const Primary = {
	args: {
		label: 'Test label',
		checked: false
	}
};
