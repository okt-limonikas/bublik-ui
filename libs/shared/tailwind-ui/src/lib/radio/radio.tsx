/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { ComponentPropsWithRef, forwardRef, SVGProps } from 'react';

import { cn } from '../utils';

const CheckedIcon = forwardRef<SVGSVGElement, SVGProps<SVGSVGElement>>(
	(props, ref) => {
		return (
			<svg
				width="12"
				height="12"
				viewBox="0 0 12 12"
				fill="none"
				xmlns="http://www.w3.org/2000/svg"
				{...props}
				ref={ref}
			>
				<rect width="12" height="12" fill="white" />
				<circle
					cx="6"
					cy="6"
					r="5.5"
					fill="currentColor"
					stroke="currentColor"
				/>
				<circle cx="6" cy="6" r="3" fill="white" />
			</svg>
		);
	}
);

const UncheckedIcon = forwardRef<SVGSVGElement, SVGProps<SVGSVGElement>>(
	(props, ref) => {
		return (
			<svg
				width="12"
				height="12"
				viewBox="0 0 12 12"
				fill="none"
				xmlns="http://www.w3.org/2000/svg"
				{...props}
				ref={ref}
			>
				<rect width="12" height="12" fill="white" />
				<circle cx="6" cy="6" r="5.5" stroke="currentColor" />
			</svg>
		);
	}
);

export interface RadioProps extends ComponentPropsWithRef<'input'> {
	label: string;
}

export const Radio = forwardRef<HTMLInputElement, RadioProps>((props, ref) => {
	const { label, ...rest } = props;

	return (
		<div data-testid="radio-field" ref={ref}>
			<label
				className={cn(
					'flex items-center w-full h-full gap-2 leading-[0.75rem] transition-all cursor-pointer text-[0.75rem] font-medium px-3 py-2 border border-border-primary rounded focus-within:shadow-[0_0_0_2px_rgba(98,126,251,0.1)]',
					props.checked && 'border-primary'
				)}
				aria-checked={props.checked}
				data-testid="tw-radio"
			>
				{props.checked ? (
					<CheckedIcon className="text-primary" />
				) : (
					<UncheckedIcon className="text-text-menu" />
				)}
				<input
					className="absolute w-0 h-0 opacity-0 appearance-none"
					type="radio"
					{...rest}
				/>
				<span className="text-[0.75rem] leading-[0.875rem]">{label}</span>
			</label>
		</div>
	);
});
