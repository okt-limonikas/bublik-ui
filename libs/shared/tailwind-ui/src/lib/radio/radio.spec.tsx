/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { it, describe, expect } from 'vitest';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Radio, RadioProps } from './radio';

const getRadioProps = (): RadioProps => {
	return { label: 'Test label' };
};

describe('Radio', () => {
	it('should render successfully', () => {
		const { baseElement } = render(<Radio {...getRadioProps()} />);

		expect(baseElement).toBeTruthy();
	});

	it('should match snapshot', () => {
		const tree = renderer.create(<Radio {...getRadioProps()} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render checked styles', () => {
		const tree = renderer
			.create(<Radio {...getRadioProps()} checked={true} />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
