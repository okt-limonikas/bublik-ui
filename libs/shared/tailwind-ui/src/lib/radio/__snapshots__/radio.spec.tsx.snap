// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Radio > should match snapshot 1`] = `
<div
  data-testid="radio-field"
>
  <label
    className="flex items-center w-full h-full gap-2 transition-all cursor-pointer text-[0.75rem] font-medium px-3 py-2 border border-border-primary rounded focus-within:shadow-[0_0_0_2px_rgba(98,126,251,0.1)]"
    data-testid="tw-radio"
  >
    <svg
      className="text-text-menu"
      fill="none"
      height="12"
      viewBox="0 0 12 12"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="white"
        height="12"
        width="12"
      />
      <circle
        cx="6"
        cy="6"
        r="5.5"
        stroke="currentColor"
      />
    </svg>
    <input
      className="absolute w-0 h-0 opacity-0 appearance-none"
      type="radio"
    />
    <span
      className="text-[0.75rem] leading-[0.875rem]"
    >
      Test label
    </span>
  </label>
</div>
`;

exports[`Radio > should render checked styles 1`] = `
<div
  data-testid="radio-field"
>
  <label
    aria-checked={true}
    className="flex items-center w-full h-full gap-2 transition-all cursor-pointer text-[0.75rem] font-medium px-3 py-2 border rounded focus-within:shadow-[0_0_0_2px_rgba(98,126,251,0.1)] border-primary"
    data-testid="tw-radio"
  >
    <svg
      className="text-primary"
      fill="none"
      height="12"
      viewBox="0 0 12 12"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="white"
        height="12"
        width="12"
      />
      <circle
        cx="6"
        cy="6"
        fill="currentColor"
        r="5.5"
        stroke="currentColor"
      />
      <circle
        cx="6"
        cy="6"
        fill="white"
        r="3"
      />
    </svg>
    <input
      checked={true}
      className="absolute w-0 h-0 opacity-0 appearance-none"
      type="radio"
    />
    <span
      className="text-[0.75rem] leading-[0.875rem]"
    >
      Test label
    </span>
  </label>
</div>
`;
